### PR TITLE
ARROW-7372: [C++] Allow creating dictionary array from simple JSON

### PR DESCRIPTION
**Pull request by pitrou**
Simple value types are supported: integers, string-like, decimal

### CODE REVIEW

**Review by bkietz — APPROVED** [review]
Very neatly done, just a few nits

File: cpp/src/arrow/ipc/json_simple.cc [hunk 1]
```diff
@@ -50,13 +51,35 @@ namespace json {
...
+template <typename Type>
+struct DictionaryBuilderTraits {
+  using BuilderType = DictionaryBuilder<Type>;
+
+  static const std::shared_ptr<DataType>& value_type(
+      const std::shared_ptr<DataType>& type) {
+    return checked_cast<const DictionaryType&>(*type).value_type();
+  }
+};
+
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/ipc/json_simple.cc:82]: Ah, right, thank you.

File: cpp/src/arrow/array/builder_dict.h [hunk 2]
```diff
@@ -190,6 +190,12 @@ class DictionaryBuilderBase : public ArrayBuilder {
   /// \brief The current number of entries in the dictionary
   int64_t dictionary_length() const { return memo_table_->size(); }
 
+  /// \brief The value byte width (for FixedSizeBinaryType)
+  template <typename T1 = T>
+  enable_if_fixed_size_binary<T1, int32_t> byte_width() const {
+    return byte_width_;
+  }
+
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/array/builder_dict.h:198]: The property is already exposed in `FixedSizeBinaryBuilder`.

**Comment by pitrou** [comment]
Thanks for the suggestions. I'm going to merge when CI is green.

**Pull request closed without merging**